File: src/main/java/leetcode/topinterview/LeetCode69.java
package leetcode.topinterview;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/22 02:10
 */
public class LeetCode69 {

    public int mySqrt(int x) {
        int l = 0, r = x, ans = -1;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if ((long) mid * mid <= x) {
                ans = mid;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return ans;
    }
}

File: src/main/go/leetcode/lists/topinterview/LeetCode69.go
package topinterview

func mySqrt(x int) int {
	left, right := 0, x
	for left <= right {
		mid := (left + right) / 2
		t := mid * mid
		if t == x {
			return mid
		}
		if t < x {
			left = mid + 1
		} else if t > x {
			right = mid - 1
		}
	}
	return right
}
